mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -220,181 +166,73 @@ struct BackpropOpPattern : public OpConversionPattern<BackpropOp> {
+        }
 
         // Create the Enzyme function
-        StringRef backpropFnName = "__enzyme_autodiff";
         Type backpropFnSignature =
             LLVM::LLVMFunctionType::get(LLVM::LLVMVoidType::get(ctx), {}, /*isVarArg=*/true);
 
         LLVM::LLVMFuncOp backpropFnDecl =
-            ensureFunctionDeclaration(rewriter, op, backpropFnName, backpropFnSignature);
-
-        // Create the memset function
-        StringRef memsetFnName = "memset";
-        Type memsetFnSignature =
-            LLVM::LLVMFunctionType::get(LLVM::LLVMPointerType::get(ctx),
-                                        {LLVM::LLVMPointerType::get(ctx), IntegerType::get(ctx, 32),
-                                         IntegerType::get(ctx, 64)});
-
-        LLVM::LLVMFuncOp memsetFnDecl =
-            ensureFunctionDeclaration(rewriter, op, memsetFnName, memsetFnSignature);
-
-        // Generate the wrapper function for argmapfn
-        func::FuncOp wrapper = genEnzymeWrapperFunction(rewriter, loc, callee);
-
-        // Set up the first argument and transforms the wrapper value in ptr
-        FunctionType wrapperType = wrapper.getFunctionType();
-        Value wrapperValue =
-            rewriter.create<func::ConstantOp>(loc, wrapperType, wrapper.getName()).getResult();
-        Type wrapperPtrType = llvmTypeConverter.convertType(wrapperType);
-        Value wrapperPtr =
-            rewriter.create<UnrealizedConversionCastOp>(loc, wrapperPtrType, wrapperValue)
-                .getResult(0);
-        // Add the pointer to the wrapped callee
-        SmallVector<Value> callArgs = {wrapperPtr};
-
-        // Get the diff arg indices
-        std::vector<size_t> diffArgIndices{0};
-        if (op.getDiffArgIndices().has_value()) {
-            auto range = op.getDiffArgIndices().value().getValues<size_t>();
-            diffArgIndices = std::vector<size_t>(range.begin(), range.end());
-        }
+            ensureFunctionDeclaration(rewriter, op, enzyme_autodiff_func_name, backpropFnSignature);
+
+        // The first argument to Enzyme is a function pointer of the function to be differentiated
+        Value calleePtr =
+            rewriter.create<func::ConstantOp>(loc, callee.getFunctionType(), callee.getName());
+        calleePtr = castToConvertedType(calleePtr, rewriter, loc);
+        SmallVector<Value> callArgs = {calleePtr};
+
+        std::vector<size_t> diffArgIndices = catalyst::compDiffArgIndices(op.getDiffArgIndices());
+        getOrInsertEnzymeGlobal(rewriter, moduleOp, enzyme_const_key);
+        getOrInsertEnzymeGlobal(rewriter, moduleOp, enzyme_dupnoneed_key);
 
         int index = 0;
         ValueRange dataIn = adaptor.getDataIn();
+
         // Add the arguments and their shadow on data in
-        for (auto [memrefArg, llvmMemrefArg] : llvm::zip(op.getArgs(), adaptor.getArgs())) {
+        for (auto [arg, llvmMemrefArg] : llvm::zip(op.getArgs(), adaptor.getArgs())) {
```

Comment:
```suggestion
        for (auto [arg, llvmArg] : llvm::zip(op.getArgs(), adaptor.getArgs())) {
```

Reply:
Ah, this `lvmMemrefArg`/`llvmArg` is no longer required because it's created via the `castToConvertedType` method.